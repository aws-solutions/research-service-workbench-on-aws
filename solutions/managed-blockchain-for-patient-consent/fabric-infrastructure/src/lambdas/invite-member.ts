/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */
import {
  ManagedBlockchainClient,
  CreateProposalCommand,
  VoteOnProposalCommand
} from '@aws-sdk/client-managedblockchain';
import { CdkCustomResourceEvent } from 'aws-lambda';

const client: ManagedBlockchainClient = new ManagedBlockchainClient({});

export const inviteMemberHandler = async (event: CdkCustomResourceEvent): Promise<void> => {
  const {
    MEMBER_ID: memberId,
    NETWORK_ID: networkId,
    MEMBERS_TO_INVITE: membersToInvite
  } = process.env as {
    MEMBER_ID: string;
    NETWORK_ID: string;
    MEMBERS_TO_INVITE: string;
  };
  if (event.RequestType === 'Create') {
    for (const accountId of membersToInvite.split(',')) {
      try {
        const proposalParams = {
          Actions: {
            Invitations: [
              {
                Principal: accountId
              }
            ]
          },
          MemberId: memberId,
          NetworkId: networkId,
          Description: 'Invite new member '
        };

        const proposalCommand = new CreateProposalCommand(proposalParams);
        const proposalResponse = await client.send(proposalCommand);

        const voteParams = {
          VoterMemberId: memberId,
          NetworkId: networkId,
          Vote: 'YES',
          ProposalId: proposalResponse.ProposalId
        };

        const command = new VoteOnProposalCommand(voteParams);
        const response = await client.send(command);
        console.log(response);
        console.log(`Successfully invited account ${accountId}`);
      } catch (error) {
        console.error(`Failed to invite account ${accountId}, encountered error ${error}`);
      }
    }
  }
};
