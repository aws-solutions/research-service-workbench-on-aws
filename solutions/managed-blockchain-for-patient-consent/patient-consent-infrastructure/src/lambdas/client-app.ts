import { Context, APIGatewayProxyResult, APIGatewayEvent } from 'aws-lambda';

export const lambdaHandler = async (): Promise<APIGatewayProxyResult> => {
  return {
    statusCode: 200,
    body: JSON.stringify({
      message: 'hello world'
    })
  };
};
