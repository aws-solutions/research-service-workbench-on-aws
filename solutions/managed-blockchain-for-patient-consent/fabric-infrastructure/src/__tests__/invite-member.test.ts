import {
  ManagedBlockchainClient,
  CreateProposalCommand,
  VoteOnProposalCommand
} from '@aws-sdk/client-managedblockchain';
import { mockClient } from 'aws-sdk-client-mock';
import { inviteMemberHandler } from '../lambdas/invite-member';

describe('inviteMemberHandler', () => {
  const ORIGINAL_ENV = process.env;
  let mockCreateProposal: jest.Mock;
  let mockVoteProposal: jest.Mock;
  beforeAll(() => {
    process.env = { ...ORIGINAL_ENV }; // Make a copy
    process.env.MEMBER_ID = 'm-somefakememberid';
    process.env.NETWORK_ID = 'n-somefakenetworkid';
    process.env.MEMBERS_TO_INVITE = '1234567890, 0987654321';
  });

  beforeEach(() => {
    mockCreateProposal = jest.fn().mockResolvedValue({ ProposalId: 'p-1234567' });
    mockVoteProposal = jest.fn().mockResolvedValue({});

    const mockBlockchainClient = mockClient(ManagedBlockchainClient);
    mockBlockchainClient.on(CreateProposalCommand).callsFake(mockCreateProposal);
    mockBlockchainClient.on(VoteOnProposalCommand).callsFake(mockVoteProposal);
  });

  afterAll(() => {
    process.env = ORIGINAL_ENV; // Restore old environment
  });
  test('Invite 2 members on RequestType Create', async () => {
    await inviteMemberHandler({
      LogicalResourceId: '',
      RequestId: '',
      RequestType: 'Create',
      ResourceProperties: { ServiceToken: '' },
      ResourceType: '',
      ResponseURL: '',
      ServiceToken: '',
      StackId: ''
    });

    expect(mockCreateProposal).toMatchSnapshot();
    expect(mockVoteProposal).toMatchSnapshot();
  });

  test('Do nothing on RequestType Update', async () => {
    await inviteMemberHandler({
      LogicalResourceId: '',
      OldResourceProperties: {},
      PhysicalResourceId: '',
      RequestId: '',
      RequestType: 'Update',
      ResourceProperties: { ServiceToken: '' },
      ResourceType: '',
      ResponseURL: '',
      ServiceToken: '',
      StackId: ''
    });

    expect(mockCreateProposal).not.toHaveBeenCalled();
    expect(mockVoteProposal).not.toHaveBeenCalled();
  });
});
