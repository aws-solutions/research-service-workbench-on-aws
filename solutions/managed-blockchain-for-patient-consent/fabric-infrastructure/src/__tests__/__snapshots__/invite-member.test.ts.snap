// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inviteMemberHandler Invite 2 members on RequestType Create 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "Actions": Object {
          "Invitations": Array [
            Object {
              "Principal": "1234567890",
            },
          ],
        },
        "Description": "Invite new member ",
        "MemberId": "m-somefakememberid",
        "NetworkId": "n-somefakenetworkid",
      },
    ],
    Array [
      Object {
        "Actions": Object {
          "Invitations": Array [
            Object {
              "Principal": " 0987654321",
            },
          ],
        },
        "Description": "Invite new member ",
        "MemberId": "m-somefakememberid",
        "NetworkId": "n-somefakenetworkid",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`inviteMemberHandler Invite 2 members on RequestType Create 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "NetworkId": "n-somefakenetworkid",
        "ProposalId": "p-1234567",
        "Vote": "YES",
        "VoterMemberId": "m-somefakememberid",
      },
    ],
    Array [
      Object {
        "NetworkId": "n-somefakenetworkid",
        "ProposalId": "p-1234567",
        "Vote": "YES",
        "VoterMemberId": "m-somefakememberid",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;
